src/class/audio/audio_device.c

```diff
@@ -1970,6 +2038,150 @@ bool audiod_xfer_cb(uint8_t rhport, uint8_t ep_addr, xfer_result_t result, uint3
+uint8_t tud_audio_n_get_fb_n_frames(uint8_t func_id)
+{
+  return _audiod_fct[func_id].fb_n_frames;
+}
+#endif
+
+#if CFG_TUD_AUDIO_ENABLE_EP_OUT && CFG_TUD_AUDIO_ENABLE_FEEDBACK_EP && (CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_OPTION == CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_OPTION_SOF_BY_AUDIO_DRIVER)
+
+// This function must be called from the user within the tud_audio_set_itf_cb(rhport, p_request) callback function, where p_request needs to be checked as
+// uint8_t const itf = tu_u16_low(p_request->wIndex);
+// uint8_t const alt = tu_u16_low(p_request->wValue);
+// such that tud_audio_set_fb_params() gets called with the parameters corresponding to the defined interface and alternate setting
+// Also, start the main clock cycle counter (or reset its value) within tud_audio_set_itf_cb()
+TU_ATTR_WEAK bool tud_audio_set_fb_params(uint8_t func_id, uint32_t f_m, uint32_t f_s, uint32_t * p_cycle_count)
+{
+  audiod_function_t* audio = &_audiod_fct[func_id];
+  audio->fb_param_p_cycle_count = p_cycle_count;
+
+  // Check if frame interval is within sane limits
+  // The interval value audio->fb_n_frames was taken from the descriptors within audiod_set_interface()
+
+  // n_frames_min is ceil(2^10 * f_s / f_m) for full speed and ceil(2^13 * f_s / f_m) for high speed - this lower limit ensures the measures feedback value has sufficient precision
+  if ((TUSB_SPEED_FULL == tud_speed_get() && ((2^10 * f_s / f_m) + 1) > audio->fb_n_frames) || (TUSB_SPEED_HIGH == tud_speed_get() && ((2^13 * f_s / f_m) + 1) > audio->fb_n_frames))
+  {
+    TU_LOG2("  UAC2 feedback interval too small\r\n"); TU_BREAKPOINT(); return false;
+  }
+
+
+#if CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE == CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE_POWER_OF_TWO_SHIFT
+  // Check if parameters really allow for a power of two division
+  if ((f_m % f_s) != 0 || !tu_is_power_of_two(audio->fb_n_frames * f_m / f_s))
+  {
+    TU_LOG2("  FEEDBACK_DETERMINATION_MODE_POWER_OF_TWO_SHIFT not possible!\r\n"); TU_BREAKPOINT(); return false;
+  }
+
+  audio->fb_power_of_two_val = 16 - tu_log2(audio->fb_n_frames * f_m / f_s);
+#endif
+
+#if CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE == CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE_FLOAT
+  audio->fb_float_val = (float)f_s / audio->fb_n_frames / f_m * (1 << 16);
+#endif
+
+#if CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE == CFG_TUD_AUDIO_ENABLE_FEEDBACK_DETERMINATION_MODE_FIXED_POINT
+
+  // f_s max is 2^19-1 = 524287 Hz
+  // f_m max is 2^29/(1 ms * n_frames) for full speed and 2^29/(125 us * n_frames) for high speed, this means for f_m fitting in an uint32, n_frames must be < 125 for full speed and < 1000 for high speed (1000 does not fit into uint8 so the maximum possible value cannot even be reached by UAC2 - we don't need to check for it)
```

Comment:
the bInterval means 2^(interval-1), though I think we almost not going more than 128 (bInterval = 8) in most cases.

![image](https://user-images.githubusercontent.com/249515/168241393-9ac156c0-9a62-4179-8644-99d20064ab7b.png)

Though I don't understand why we don't just shift f_s by 16 then multiply with cycle later on instead of shift 13 on f_s and 3 on cycle ?

Reply:
That is correct, my original intention was to save N and D as unit32 in order to save space. 13/3 was chosen because i wanted f_s to be able to hold frequencies of bigger than 500 kHz (just to cover a large space also for application other than audio). And i wanted the scale on n_cycles to be a small as possible because n_cycle must cover the same range as f_m, for which we need a huge number. I think 8/8 would be not a good choice therefor.
During rewriting the code, i realized that N * n_cylces must be unit64, so i ended up extending everything. Now, you are correct, we can simple choose to scale N as N << 16 and skip the << 3 later on! I forgot about that, well spotted!

Reply:
thanks for confirmation, I have an update proposal for this PR detailed below :)